controllers/get_resource.go

```diff
@@ -5,72 +5,41 @@
 package controllers
 
 import (
-	"context"
 	"errors"
 	"fmt"
 	"io"
-	"net/http"
 	"net/url"
 
-	"github.com/google/go-containerregistry/pkg/name"
-	"github.com/google/go-containerregistry/pkg/v1/remote"
 	"github.com/open-component-model/ocm/pkg/contexts/ocm"
 	"github.com/open-component-model/ocm/pkg/contexts/ocm/accessmethods/localblob"
 	"github.com/open-component-model/ocm/pkg/contexts/ocm/accessmethods/ociartefact"
 	"github.com/open-component-model/ocm/pkg/contexts/ocm/accessmethods/ociblob"
 	ocmapi "github.com/open-component-model/ocm/pkg/contexts/ocm/compdesc/versions/ocm.software/v3alpha1"
 	"github.com/open-component-model/ocm/pkg/contexts/ocm/cpi"
 	ocmruntime "github.com/open-component-model/ocm/pkg/runtime"
-)
-
-func GetResource(ctx context.Context, ociRegistryAddress string, resource *ocmapi.Resource, result interface{}) error {
-	access, err := GetImageReference(resource)
-	if err != nil {
-		return fmt.Errorf("failed to create digest: %w", err)
-	}
 
-	digest, err := name.NewDigest(access, name.Insecure)
-	if err != nil {
-		return fmt.Errorf("failed to create digest: %w", err)
-	}
+	"github.com/open-component-model/ocm-controller/api/v1alpha1"
+	"github.com/open-component-model/ocm-controller/pkg/oci"
+)
 
-	// proxy image requests via the in-cluster oci-registry
-	proxyURL, err := url.Parse(fmt.Sprintf("http://%s", ociRegistryAddress))
+func GetResource(snapshot v1alpha1.Snapshot, result interface{}) error {
+	u, err := url.Parse(snapshot.Status.Image)
 	if err != nil {
-		return fmt.Errorf("failed to parse oci registry url: %w", err)
+		return fmt.Errorf("failed to parse URL: %w", err)
 	}
-
-	// create transport to the in-cluster oci-registry
-	tr := newCustomTransport(remote.DefaultTransport.(*http.Transport).Clone(), proxyURL)
-
-	// set context values to be transmitted as headers on the registry requests
-	for k, v := range map[string]string{
-		"registry":   digest.Repository.Registry.String(),
-		"repository": digest.Repository.String(),
-		"digest":     digest.String(),
-		"image":      digest.Name(),
-		"tag":        resource.Version,
-	} {
-		ctx = context.WithValue(ctx, contextKey(k), v)
-	}
-
-	// fetch the layer
-	remoteLayer, err := remote.Layer(digest, remote.WithTransport(tr), remote.WithContext(ctx))
+	repo, err := oci.NewRepository(u.Host, oci.WithInsecure())
 	if err != nil {
-		return fmt.Errorf("failed to retrieve layer: %w", err)
+		return fmt.Errorf("failed to get repository: %w", err)
 	}
-
-	data, err := remoteLayer.Uncompressed()
+	blob, err := repo.FetchBlob(snapshot.Spec.Digest)
```

Comment:
for me `snapshot.Status` should have everything needed to retrieve an existing artifact. `.Spec` should be used only for new/non-existing one.